File: src/cn/vove7/plugin/rest/filetype/RestCodeStyleSettingsProvider.java
package cn.vove7.plugin.rest.filetype;

import com.intellij.application.options.CodeStyleAbstractConfigurable;
import com.intellij.application.options.CodeStyleAbstractPanel;
import com.intellij.application.options.TabbedLanguageCodeStylePanel;
import com.intellij.openapi.options.Configurable;
import com.intellij.psi.codeStyle.CodeStyleSettings;
import com.intellij.psi.codeStyle.CodeStyleSettingsProvider;
import com.intellij.psi.codeStyle.CustomCodeStyleSettings;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * @author danblack
 */
public class RestCodeStyleSettingsProvider extends CodeStyleSettingsProvider {
    @NotNull
    @Override
    public Configurable createSettingsPage(CodeStyleSettings settings, CodeStyleSettings originalSettings) {
        return new CodeStyleAbstractConfigurable(settings, originalSettings, "Rest") {

            @Nullable
            @Override
            public String getHelpTopic() {
                return null;
            }

            @Override
            protected CodeStyleAbstractPanel createPanel(CodeStyleSettings settings) {
                return new RestTabbedLanguageCodeStylePanel(getCurrentSettings(), settings);
            }
        };
    }

    @Nullable
    @Override
    public String getConfigurableDisplayName() {
        return "Rest";
    }

    @Nullable
    @Override
    public CustomCodeStyleSettings createCustomSettings(CodeStyleSettings settings) {
        return new RestCodeStyleSettings(settings);
    }

    private static class RestTabbedLanguageCodeStylePanel extends TabbedLanguageCodeStylePanel {

        protected RestTabbedLanguageCodeStylePanel(CodeStyleSettings currentSettings, CodeStyleSettings settings) {
            super(RestLanguage.INSTANCE, currentSettings, settings);
        }
    }
}

File: src/main/kotlin/cn/vove7/plugin/rest/filetype/RestCodeStyleSettingsProvider.kt
package cn.vove7.plugin.rest.filetype

import com.intellij.psi.codeStyle.CodeStyleSettingsProvider
import com.intellij.psi.codeStyle.CodeStyleSettings
import com.intellij.application.options.CodeStyleAbstractConfigurable
import com.intellij.application.options.CodeStyleAbstractPanel
import cn.vove7.plugin.rest.filetype.RestCodeStyleSettingsProvider.RestTabbedLanguageCodeStylePanel
import com.intellij.psi.codeStyle.CustomCodeStyleSettings
import cn.vove7.plugin.rest.filetype.RestCodeStyleSettings
import com.intellij.application.options.TabbedLanguageCodeStylePanel
import cn.vove7.plugin.rest.filetype.RestLanguage
import com.intellij.openapi.options.Configurable

/**
 * @author danblack
 */
class RestCodeStyleSettingsProvider : CodeStyleSettingsProvider() {
    override fun createSettingsPage(settings: CodeStyleSettings, originalSettings: CodeStyleSettings): Configurable {
        return object : CodeStyleAbstractConfigurable(settings, originalSettings, "Rest") {
            override fun getHelpTopic(): String? {
                return null
            }

            override fun createPanel(settings: CodeStyleSettings): CodeStyleAbstractPanel {
                return RestTabbedLanguageCodeStylePanel(currentSettings, settings)
            }
        }
    }

    override fun getConfigurableDisplayName(): String? {
        return "Rest"
    }

    override fun createCustomSettings(settings: CodeStyleSettings): CustomCodeStyleSettings? {
        return RestCodeStyleSettings(settings)
    }

    private class RestTabbedLanguageCodeStylePanel(currentSettings: CodeStyleSettings?, settings: CodeStyleSettings?) :
        TabbedLanguageCodeStylePanel(RestLanguage, currentSettings, settings)
}